### PR TITLE
Add support for custom-request authentication

### PR DESCRIPTION
**Pull request by barweiss**

### CODE REVIEW

File: seekret/apitest/_createleaf.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+def create_leaf(path: str, value):
```
Comment (asdftamir) [inline seekret/apitest/_createleaf.py]: This doesnt support arrays. Should it?

File: seekret/apitest/hooks/__init__.py [hunk 2]
```diff
@@ -19,5 +21,27 @@ def pytest_tavern_beta_after_every_response(expected, response: Response):
     :note: This is a hook function intended to be imported in a conftest.py file in the tavern test directory.
     """
 
-    print(f'Incoming response: {response.status_code} {response.reason} from {response.url}')
+    print(
```
Comment (asdftamir) [inline seekret/apitest/hooks/__init__.py:24]: didnt slavin add the method here?

File: seekret/apitest/hooks/__init__.py [hunk 3]
```diff
@@ -19,5 +21,27 @@ def pytest_tavern_beta_after_every_response(expected, response: Response):
...
+
+    test_box = Box(test_dict)
+    variables_box = Box(variables)
+
+    auth_type = variables_box.seekret.v1.users.user.auth.type
+    auth_data = variables_box.seekret.v1.users.user.auth.data
+
+    call_on_test_start(test_box, variables_box, auth_type, auth_data)
+
+    variables.clear()
```
Comment (asdftamir) [inline seekret/apitest/hooks/__init__.py]: y?

File: seekret/apitest/hooks/__init__.py [hunk 4]
```diff
@@ -19,5 +21,27 @@ def pytest_tavern_beta_after_every_response(expected, response: Response):
...
+    variables.update(variables_box.to_dict())
+
+    test_dict.clear()
```
Comment (asdftamir) [inline seekret/apitest/hooks/__init__.py]: y?

**Review by asdftamir — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -86,3 +85,51 @@ variables:
...
+      # ...
+      users:
+        user:
+          auth:
+            type: custom-request
+            data:
+              auth_stage_id: my-custom-auth-stage # Optional, defaults to "auth"
+
+stages:
+  - id: my-custom-auth-stage
```
Comment (asdftamir) [inline README.md:110]: Add an exmaple of taking from json body and using as bearer

**Review by asdftamir — CHANGES_REQUESTED** [review]

File: seekret/apitest/hooks/__init__.py (hunk 2 above)
Comment (barweiss) [reply seekret/apitest/hooks/__init__.py:24]: Nope, just content

File: seekret/apitest/_createleaf.py (hunk 1 above)
Comment (barweiss) [reply seekret/apitest/_createleaf.py]: I removed the support for now, I think it's not required for now.
We can add support when the need arises.

File: seekret/apitest/hooks/__init__.py (hunk 4 above)
Comment (barweiss) [reply seekret/apitest/hooks/__init__.py]: Clarify

File: seekret/apitest/hooks/__init__.py (hunk 3 above)
Comment (barweiss) [reply seekret/apitest/hooks/__init__.py]: Clarify

File: seekret/apitest/hooks/__init__.py (hunk 2 above)
Comment (asdftamir) [reply seekret/apitest/hooks/__init__.py:24]: so add the method (beforte the url)

File: seekret/apitest/hooks/__init__.py (hunk 3 above)
Comment (asdftamir) [reply seekret/apitest/hooks/__init__.py]: why do u clear the values?

File: seekret/apitest/hooks/__init__.py (hunk 4 above)
Comment (asdftamir) [reply seekret/apitest/hooks/__init__.py]: y do u clear the dict?

File: README.md [hunk 6]
```diff
@@ -114,22 +114,76 @@ stages:
...
-          function: seekret.apitest:dots_save
+          function: seekret.apitest:save_authorization
           extra_kwargs:
-            headers:
-              # This will take the X-Auth-Token header from the response and put it in the header
-              # of future test stages.
-              seekret-runtime.v1.auth.headers.X-Auth-Token: '"X-Auth-Token"'
+            # This will take the X-Auth-Token header from the response and put it in the header
+            # of future test stages.
+            headers: '"X-Auth-Token"'
```
Comment (asdftamir) [inline README.md]: is the double quotes necessary?

File: README.md [hunk 7]
```diff
@@ -114,22 +114,76 @@ stages:
...
+|---------------|---------------------------------------------------------------------------------------------------------|
+| `type`        | Type of authorization method (currently supported: `header`, `bearer`).                                 |
+| `data`        | Additional data for the authorization method. See the next table for expected values.                   |
+| `headers`     | JMES path to the response header to take the authorization token from. Can't be used with `json`.       |
+| `json`        | JMES path to the authorization token in the JSON content of the response. Can't be used with `headers`. |
+
+| `type` value | `data` fields                                                              |
+|--------------|----------------------------------------------------------------------------|
+| `header`     | `target_header` - the name of the header to set to the authorization token |
+| `bearer`     | N/A                                                                        |
```
Comment (asdftamir) [inline README.md]: porque bearer is not available?

Comment (asdftamir) [reply README.md]: oh i saw in the example. clarify in this table that `data` is not required

File: seekret/apitest/auth/saveutil.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        # ...
+        save:
+            $ext:
+                function: seekret.apitest:save_authorization
+                extra_kwargs:
+                    type: bearer
+                    json: data.token  # JMES path to the token.
+    """
+
+    if (not headers and json is None) or (headers and json is not None):
```
Comment (asdftamir) [inline seekret/apitest/auth/saveutil.py]: please demorgan this somehow

**Review by asdftamir — CHANGES_REQUESTED** [review]

File: seekret/apitest/hooks/__init__.py (hunk 2 above)
Comment (barweiss) [reply seekret/apitest/hooks/__init__.py:24]: Ignoring this

File: seekret/apitest/hooks/__init__.py (hunk 4 above)
Comment (barweiss) [reply seekret/apitest/hooks/__init__.py]: Using `update` now

File: seekret/apitest/hooks/__init__.py (hunk 3 above)
Comment (barweiss) [reply seekret/apitest/hooks/__init__.py]: Using `update` now

File: README.md (hunk 6 above)
Comment (barweiss) [reply README.md]: Yep, to form a JMESpath expression

File: README.md (hunk 7 above)
Comment (barweiss) [reply README.md]: Fixed

File: seekret/apitest/auth/saveutil.py (hunk 8 above)
Comment (barweiss) [reply seekret/apitest/auth/saveutil.py]: Hopefully made easier to read

File: README.md (hunk 6 above)
Comment (asdftamir) [reply README.md]: so emphasize in documentation

**Review by asdftamir — DISMISSED** [review]

**Review by asdftamir — APPROVED** [review]

**Pull request merged**